pr/amdgpu closed world

jdoerfert opened the pull request.

JonChesterfield commented:
This looks useful - did it make it in some other form?

arsenm commented:
> This looks useful - did it make it in some other form?

No, this is waiting for #83131, which is in turn waiting for some variant of #93526 or #86012

jdoerfert commented:
> > This looks useful - did it make it in some other form?
> 
> No, this is waiting for #83131, which is in turn waiting for some variant of #93526 or #86012

FWIW, it waits for the other PR, then needs an update to use the pass manager intrinsic knowledge of a post-link time world rather than the flag.
The logic itself should be ready to merge though.

jdoerfert commented:
@arsenm I rebased this, and moved the pass for now into the LTO pipeline.
That breaks tests since they run the non-LTO one to check for the Attributor results.
If I run it in the non-LTO pipeline, it only knows about closed world if you do thin-LTO, which is not ideal. Maybe there is a pipeline/extension point I missed?
WDYT?

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUTargetMachine.cpp:
```diff
@@ -731,16 +731,11 @@ void AMDGPUTargetMachine::registerPassBuilderCallbacks(PassBuilder &PB) {
       [this](ModulePassManager &PM, OptimizationLevel Level) {
+        if (Level != OptimizationLevel::O0)
+          PM.addPass(
+              AMDGPUAttributorPass(*this, /*HasWholeProgramVisibility*/ true));
```
```suggestion
              AMDGPUAttributorPass(*this, /*HasWholeProgramVisibility=*/ true));
```

arsenm commented on llvm/test/CodeGen/AMDGPU/GlobalISel/irtranslator-indirect-call.ll:
```diff
@@ -1,74 +1,75 @@
 ; NOTE: Assertions have been autogenerated by utils/update_mir_test_checks.py
-; RUN: llc -global-isel -stop-after=irtranslator -mtriple=amdgcn-mesa-mesa3d -mcpu=gfx900 -verify-machineinstrs -o - %s | FileCheck -enable-var-scope %s
+; RUN: llc -global-isel -stop-after=irtranslator -mtriple=amdgcn-mesa-mesa3d -mcpu=gfx900 -verify-machineinstrs -o - %s | FileCheck -enable-var-scope --check-prefixes=SAMEC,CHECK %s
+; RUN: llc -global-isel -stop-after=irtranslator -attributor-assume-closed-world -mtriple=amdgcn-mesa-mesa3d -mcpu=gfx900 -verify-machineinstrs -o - %s | FileCheck -enable-var-scope --check-prefixes=SAMEC,CWRLD %s
```
Why is this changing? The attributor no longer runs in the codegen pipeline

arsenm replied:
Same question

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp:
```diff
@@ -1036,14 +1041,33 @@ static bool runImpl(Module &M, AnalysisGetter &AG, TargetMachine &TM) {
   AC.Allowed = &Allowed;
   AC.IsModulePass = true;
   AC.DefaultInitializeLiveInternals = false;
+  errs() << "HasWholeProgramVisibility " << HasWholeProgramVisibility << "\n";
```
Leftover debug printing

arsenm commented on llvm/include/llvm/Transforms/IPO/Attributor.h:
```diff
@@ -1448,7 +1448,7 @@ struct AttributorConfig {
   /// direct call targets (with an if-cascade).
   std::function<bool(Attributor &A, const AbstractAttribute &AA, CallBase &CB,
-                     Function &AssummedCallee)>
+                     Function &AssummedCallee, unsigned NumCallees)>
```
Typo AssummedCallee

arsenm commented:
> @arsenm I rebased this, and moved the pass for now into the LTO pipeline. That breaks tests since they run the non-LTO one to check for the Attributor results. If I run it in the non-LTO pipeline, it only knows about closed world if you do thin-LTO, which is not ideal. Maybe there is a pipeline/extension point I missed? WDYT?

I'm confused, none of the codegen tests should be changing? LTO/Thin-LTO is upstream of all of the backend testing?

jdoerfert commented:
> > @arsenm I rebased this, and moved the pass for now into the LTO pipeline. That breaks tests since they run the non-LTO one to check for the Attributor results. If I run it in the non-LTO pipeline, it only knows about closed world if you do thin-LTO, which is not ideal. Maybe there is a pipeline/extension point I missed? WDYT?
> 
> I'm confused, none of the codegen tests should be changing? LTO/Thin-LTO is upstream of all of the backend testing?

So, when I put it to the current location (FullLTO last callback), all opt and llc test failed because they don't run that. I did not update tests yet. The other FullLTO pass uses llvm-lto2 run to verify it runs properly. The pass is executed if I do a full compile with full or thin LTO.

arsenm commented:
> So, when I put it to the current location (FullLTO last callback), all opt and llc test failed because they don't run that. 

I'm still confused. They don't run it, correct. I expect them to still not run AMDGPUAttributor after this patch (keeping all the test updates I did in b1bcb7ca460fcd317bbc8309e14c8761bf8394e0 intact.

jdoerfert commented:
> > So, when I put it to the current location (FullLTO last callback), all opt and llc test failed because they don't run that.
> 
> I'm still confused. They don't run it, correct. I expect them to still not run AMDGPUAttributor after this patch (keeping all the test updates I did in [b1bcb7c](https://github.com/llvm/llvm-project/commit/b1bcb7ca460fcd317bbc8309e14c8761bf8394e0) intact.

Options, as far as I can tell:
0) Old PM pipeline, as shown with the first version of this patch: runs always, requires the pipeline to set the closed world assumptions, only then works in a compile job with clang or opt.
1) New PM pipeline, as it is w/o this patch: runs always, never have closed world assumptions in a compile job with clang.
2) New PM pipeline, read LTOMode in the new PM to determine Post(Thin|Full)LTOLink and enable closed world assumption: runs always, has closed world assumption in ThinLTO compile only (when executed via clang), some test might work with opt.
3) New PM pipeline, add to FullLTO pipeline, as it is w/ this patch: runs always with clang (FullLTO or ThinLTO, hence all compiles), doesn't run with opt but tests need llvm-lto(2)

We could add "-attributor-assume-closed-world" to tests, as I did before, but that only works with 1) and 2). It also again requires us to set the flag from the driver/tooling (clang or llc). That's what I had before.

@shiltian offered to look into this. Maybe he has more luck than me.

arsenm commented:
> > > So, when I put it to the current location (FullLTO last callback), all opt and llc test failed because they don't run that.
> > 
> > 
> > I'm still confused. They don't run it, correct. I expect them to still not run AMDGPUAttributor after this patch (keeping all the test updates I did in [b1bcb7c](https://github.com/llvm/llvm-project/commit/b1bcb7ca460fcd317bbc8309e14c8761bf8394e0) intact.
> 
> Options, as far as I can tell: 0) Old PM pipeline, as shown with the first version of this patch: runs always, requires the pipeline to set the closed world assumptions, only then works in a compile job with clang or opt.
> 
> 1. New PM pipeline, as it is w/o this patch: runs always, never have closed world assumptions in a compile job with clang.
> 2. New PM pipeline, read LTOMode in the new PM to determine Post(Thin|Full)LTOLink and enable closed world assumption: runs always, has closed world assumption in ThinLTO compile only (when executed via clang), some test might work with opt.
> 3. New PM pipeline, add to FullLTO pipeline, as it is w/ this patch: runs always with clang (FullLTO or ThinLTO, hence all compiles), doesn't run with opt but tests need llvm-lto(2)
> 
> We could add "-attributor-assume-closed-world" to tests, as I did before, but that only works with 1) and 2). It also again requires us to set the flag from the driver/tooling (clang or llc). That's what I had before.
> 
> @shiltian offered to look into this. Maybe he has more luck than me.

I expect 2? I expect no change in llc behavior (i.e. AMDGPUAttributor does not run in llc, and thus this flag has no effect)

gandhi56 commented:
Hi @jdoerfert, thanks for working on this! I tried this patch on the following indirect call test but it seems to fail on AMDGPU, in spite of passing on X86:

```
; RUN: opt -S -mtriple=x86_64-unknown-linux-gnu -passes=attributor -attributor-assume-closed-world -o - %s | FileCheck %s --check-prefix=GENERIC
; RUN: opt -S -mtriple=amdgcn-amd-amdhsa -passes=amdgpu-attributor -attributor-assume-closed-world -o - %s | FileCheck %s --check-prefix=AMDGPU

; Function Attrs: mustprogress nounwind uwtable
define dso_local noundef i32 @_Z3foov() #0 {
; GENERIC-LABEL: define dso_local noundef i32 @_Z3foov(
; GENERIC-SAME: ) #[[ATTR0:[0-9]+]] {
; GENERIC-NEXT:    ret i32 1
;
; AMDGPU-LABEL: define dso_local noundef i32 @_Z3foov(
; AMDGPU-SAME: ) #[[ATTR0:[0-9]+]] {
; AMDGPU-NEXT:    ret i32 1
;
  ret i32 1
}

; Function Attrs: mustprogress nounwind uwtable
define dso_local noundef i32 @_Z3barv() #0 {
; GENERIC-LABEL: define dso_local noundef i32 @_Z3barv(
; GENERIC-SAME: ) #[[ATTR0]] {
; GENERIC-NEXT:    ret i32 2
;
; AMDGPU-LABEL: define dso_local noundef i32 @_Z3barv(
; AMDGPU-SAME: ) #[[ATTR0]] {
; AMDGPU-NEXT:    ret i32 2
;
  ret i32 2
}

; Function Attrs: mustprogress norecurse uwtable
define amdgpu_kernel void @main(i32 noundef %0, ptr %out) #1 {
; GENERIC-LABEL: define amdgpu_kernel void @main(
; GENERIC:    [[TMP11:%.*]] = call noundef i32 @_Z3barv()
; GENERIC:    [[TMP14:%.*]] = call noundef i32 @_Z3foov()
;
; AMDGPU-LABEL: define amdgpu_kernel void @main(
; AMDGPU:    [[TMP11:%.*]] = call noundef i32 @_Z3barv()
; AMDGPU:    [[TMP14:%.*]] = call noundef i32 @_Z3foov()
;
  %2 = alloca i32, align 4
  %3 = alloca i32, align 4
  %4 = alloca ptr, align 8
  store i32 0, ptr %2, align 4
  store i32 %0, ptr %3, align 4
  call void @llvm.lifetime.start.p0(i64 8, ptr %4) #3
  store ptr @_Z3foov, ptr %4, align 8
  %5 = load i32, ptr %3, align 4
  %6 = icmp ne i32 %5, 0
  br i1 %6, label %7, label %8

7:                                                ; preds = %1
  store ptr @_Z3barv, ptr %4, align 8
  br label %8

8:                                                ; preds = %7, %1
  %9 = load ptr, ptr %4, align 8
  %10 = call noundef i32 %9()
  store i32 %10, ptr %out, align 4
  call void @llvm.lifetime.end.p0(i64 8, ptr %4) #3
  ret void
}

; Function Attrs: nocallback nofree nosync nounwind willreturn memory(argmem: readwrite)
declare void @llvm.lifetime.start.p0(i64 immarg, ptr nocapture) #2

; Function Attrs: nocallback nofree nosync nounwind willreturn memory(argmem: readwrite)
declare void @llvm.lifetime.end.p0(i64 immarg, ptr nocapture) #2

attributes #0 = { mustprogress nounwind uwtable }
attributes #1 = { mustprogress norecurse uwtable }
attributes #2 = { nocallback nofree nosync nounwind willreturn memory(argmem: readwrite) }
attributes #3 = { nounwind }
```

arsenm commented:
> Hi @jdoerfert, thanks for working on this! I tried this patch on the following indirect call test but it seems to fail on AMDGPU, in spite of passing on X86:

What do you mean by fail? What does it fail to do?

gandhi56 commented:
> > Hi @jdoerfert, thanks for working on this! I tried this patch on the following indirect call test but it seems to fail on AMDGPU, in spite of passing on X86:
> 
> What do you mean by fail? What does it fail to do?

It fails to specialize the indirect call into two direct calls. Never mind though, I suppose the specialization would happen later in the pipeline.

shiltian commented:
I'll carry on this PR.

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp:
```diff
@@ -1036,14 +1041,32 @@ static bool runImpl(Module &M, AnalysisGetter &AG, TargetMachine &TM) {
+        // Singleton functions should be specialized.
+        if (NumCallees == 1)
+          return true;
+        // Otherewise specialize uniform values.
```
Typo Otherewise

arsenm commented on llvm/lib/Target/AMDGPU/AMDGPUAttributor.cpp:
```diff
@@ -1036,14 +1041,32 @@ static bool runImpl(Module &M, AnalysisGetter &AG, TargetMachine &TM) {
+        if (NumCallees == 1)
+          return true;
+        // Otherewise specialize uniform values.
+        const auto &TTI = TM.getTargetTransformInfo(*CB.getCaller());
```
It's unfortunate this non-subtarget dependent property requires you to query the function's subtarget

jdoerfert committed:
[Attributor][NFC] Add NumCallees argument to callback

jdoerfert committed:
[Attributor][AMDGPU] Improve indirect call support in closed modules

If we see all functions that can be called, thus in a "closed world",
we can perform better reasoning in the presence of unknown callees of
indirect calls. We now collect all indirectly callable functions and
limit the potentially called functions to those.

The AMDGPU backend is the only user for now. We should enable this for
AMDGPU (and NVIDIA GPUs in certain cases) also when we run the
Attributor (or OpenMP-opt) earlier in the pipeline.

shiltian committed:
Take over and fix comments

shiltian commented:
Now I have removed those changes in tests for `llc`, as `AMDGPUAttributorPass` is moved to middle end. One of the major changes in this PR is to again move `AMDGPUAttributorPass` to LTO. If we decide to do so, the movement needs to be stacked before this PR.

shiltian commented:
> Hi @jdoerfert, thanks for working on this! I tried this patch on the following indirect call test but it seems to fail on AMDGPU, in spite of passing on X86:
> 
> ```
> ; RUN: opt -S -mtriple=x86_64-unknown-linux-gnu -passes=attributor -attributor-assume-closed-world -o - %s | FileCheck %s --check-prefix=GENERIC
> ; RUN: opt -S -mtriple=amdgcn-amd-amdhsa -passes=amdgpu-attributor -attributor-assume-closed-world -o - %s | FileCheck %s --check-prefix=AMDGPU
> 
> ; Function Attrs: mustprogress nounwind uwtable
> define dso_local noundef i32 @_Z3foov() #0 {
> ; GENERIC-LABEL: define dso_local noundef i32 @_Z3foov(
> ; GENERIC-SAME: ) #[[ATTR0:[0-9]+]] {
> ; GENERIC-NEXT:    ret i32 1
> ;
> ; AMDGPU-LABEL: define dso_local noundef i32 @_Z3foov(
> ; AMDGPU-SAME: ) #[[ATTR0:[0-9]+]] {
> ; AMDGPU-NEXT:    ret i32 1
> ;
>   ret i32 1
> }
> 
> ; Function Attrs: mustprogress nounwind uwtable
> define dso_local noundef i32 @_Z3barv() #0 {
> ; GENERIC-LABEL: define dso_local noundef i32 @_Z3barv(
> ; GENERIC-SAME: ) #[[ATTR0]] {
> ; GENERIC-NEXT:    ret i32 2
> ;
> ; AMDGPU-LABEL: define dso_local noundef i32 @_Z3barv(
> ; AMDGPU-SAME: ) #[[ATTR0]] {
> ; AMDGPU-NEXT:    ret i32 2
> ;
>   ret i32 2
> }
> 
> ; Function Attrs: mustprogress norecurse uwtable
> define amdgpu_kernel void @main(i32 noundef %0, ptr %out) #1 {
> ; GENERIC-LABEL: define amdgpu_kernel void @main(
> ; GENERIC:    [[TMP11:%.*]] = call noundef i32 @_Z3barv()
> ; GENERIC:    [[TMP14:%.*]] = call noundef i32 @_Z3foov()
> ;
> ; AMDGPU-LABEL: define amdgpu_kernel void @main(
> ; AMDGPU:    [[TMP11:%.*]] = call noundef i32 @_Z3barv()
> ; AMDGPU:    [[TMP14:%.*]] = call noundef i32 @_Z3foov()
> ;
>   %2 = alloca i32, align 4
>   %3 = alloca i32, align 4
>   %4 = alloca ptr, align 8
>   store i32 0, ptr %2, align 4
>   store i32 %0, ptr %3, align 4
>   call void @llvm.lifetime.start.p0(i64 8, ptr %4) #3
>   store ptr @_Z3foov, ptr %4, align 8
>   %5 = load i32, ptr %3, align 4
>   %6 = icmp ne i32 %5, 0
>   br i1 %6, label %7, label %8
> 
> 7:                                                ; preds = %1
>   store ptr @_Z3barv, ptr %4, align 8
>   br label %8
> 
> 8:                                                ; preds = %7, %1
>   %9 = load ptr, ptr %4, align 8
>   %10 = call noundef i32 %9()
>   store i32 %10, ptr %out, align 4
>   call void @llvm.lifetime.end.p0(i64 8, ptr %4) #3
>   ret void
> }
> 
> ; Function Attrs: nocallback nofree nosync nounwind willreturn memory(argmem: readwrite)
> declare void @llvm.lifetime.start.p0(i64 immarg, ptr nocapture) #2
> 
> ; Function Attrs: nocallback nofree nosync nounwind willreturn memory(argmem: readwrite)
> declare void @llvm.lifetime.end.p0(i64 immarg, ptr nocapture) #2
> 
> attributes #0 = { mustprogress nounwind uwtable }
> attributes #1 = { mustprogress norecurse uwtable }
> attributes #2 = { nocallback nofree nosync nounwind willreturn memory(argmem: readwrite) }
> attributes #3 = { nounwind }
> ```

It can't be specialized because the function pointer is not uniform.

shiltian commented:
We can close this issue for now and discuss further in https://github.com/llvm/llvm-project/pull/100953 and https://github.com/llvm/llvm-project/pull/102086.
